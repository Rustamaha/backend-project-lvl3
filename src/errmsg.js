const errno = require('errno');

const errmsg = (err) => {
  let str = 'Error: ';
  if (errno.errno[err.errno]) {
    str += errno.errno[err.errno].description;
  } else {
    str += err.message;
  }
  if (err.path) {
    str += ` [ ${err.path} ]`;
  }
  return str;
};

export default errmsg;
