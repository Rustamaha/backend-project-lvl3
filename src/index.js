import pageLoader from './loader';

export default pageLoader;
