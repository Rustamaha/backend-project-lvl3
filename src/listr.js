import path from 'path';
import { promises as fs } from 'fs';

const Listr = require('listr');

const loadAndSaveData = (promises, paths, pathToFiles) => {
  const taskList = promises.map(request => new Listr([
    {
      title: '',
      task: (ctx, taskObj) => request
        .then((response) => {
          const { url } = response.config;
          taskObj.title = `downloading ${url}`; //  eslint-disable-line
          const { data } = response;
          const { fileName } = paths[url];
          return fs.writeFile(path.resolve(pathToFiles, fileName), data);
        })
        .catch((err) => {
          ctx.error = err;
          taskObj.title = err.message; //  eslint-disable-line
        }),
    },
  ], [{ concurrent: true }]).run({}));
  return Promise.all(taskList);
};

export default loadAndSaveData;
