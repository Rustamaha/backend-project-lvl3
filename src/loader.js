import axios from 'axios';
import cheerio from 'cheerio';
import httpAdapter from 'axios/lib/adapters/http';
import path from 'path';
import { promises as fs } from 'fs';

import loadAndSaveData from './listr';

const debug = require('debug');

const log = debug('page-loader:loader');

axios.defaults.adapter = httpAdapter;

const currentDir = process.cwd();
log(`Current directory: ${currentDir}`);

export const makeName = (url, type) => {
  const reProtocol = /^https:\/\/|^\/\/|^\//;
  const re = /[\.,/,?,=]/g; //  eslint-disable-line
  const name = url.replace(reProtocol, '').replace(re, '-');
  return type ? `${name}${type}` : `${name}`;
};

const pageLoader = (pageUrl, pathDir = currentDir) => {
  const stats = fs.stat(pathDir);
  const {
    hostname,
    origin,
  } = new URL(pageUrl);
  const reSld = /(\.[^.]*)(\.*$)/;
  const secLvlDomain = hostname.replace(reSld, '').split('.').pop();

  let page;
  const paths = {};

  const dirName = makeName(pageUrl, '_files');
  const request = stats.then(() => axios.get(pageUrl))
    .then(({ data }) => {
      page = data;
      return data;
    });

  const dom = request.then((data) => {
    const $ = cheerio.load(data);
    return $;
  });

  const filter = ($, attr) => (i, el) => {
    const pathEl = $(el).attr(attr);
    const reUrl = /^https:\/\/[a-z.0-9-]{2,}\.[a-z]{2,}|^\/\/[a-z.0-9-]{2,}\.[a-z]{2,}/;
    const reCutUrl = /^\/[a-z0-9-]{2,}/;
    if (pathEl && pathEl.match(reUrl)) {
      const matched = pathEl.match(reUrl);
      const hostName = matched[0].slice(2).split('.');
      const secLvlD = hostName[hostName.length - 2];
      return secLvlD === secLvlDomain;
    }
    return pathEl && pathEl.match(reCutUrl);
  };

  const traverse = ($, attr) => (i, el) => {
    const pathEl = $(el).attr(attr);
    const reUrlWithHttps = /^https:\/\/[a-z.0-9-]{2,}\.[a-z]{2,}/;
    const reUrlWithoutHttps = /^\/\/[a-z.0-9-]{2,}\.[a-z]{2,}/;
    const reType = /\.[a-z]{2,}$|\.[a-z]{2,}\/[a-z0-9]{2,}$/g;
    const typeMatch = pathEl.match(reType);
    let type;
    if (typeMatch) {
      const [ending] = typeMatch[0].split('/');
      type = ending;
    }
    const fileName = makeName(pathEl, type);
    $(el).attr({
      src: `./${dirName}/${fileName}`,
    });
    page = $.html();
    if (pathEl.match(reUrlWithHttps)) {
      paths[pathEl] = { fileName, type };
      return;
    }
    if (pathEl.match(reUrlWithoutHttps)) {
      const makeUrl = `https:${pathEl}`;
      paths[makeUrl] = { fileName, type };
      return;
    }
    const makeUrl = `${origin}${pathEl}`;
    paths[makeUrl] = { fileName, type };
  };

  const scripts = dom.then($ => $('script').filter(filter($, 'src'))
    .map(traverse($, 'src')));
  const links = dom.then($ => $('link').filter(filter($, 'href'))
    .map(traverse($, 'href')));
  const images = dom.then($ => $('img').filter(filter($, 'src'))
    .map(traverse($, 'src')));

  let promises;
  const html = Promise.all([scripts, links, images])
    .then(() => {
      const getData = Object.keys(paths).map((key) => {
        const { type } = paths[key];
        if (type === '.png' || type === '.svg' || type === '.jpeg' || type === '.jpg') {
          return axios.get(key, { responseType: 'arraybuffer' });
        }
        return axios.get(key);
      });
      promises = getData;
    });

  const pageName = makeName(pageUrl, '.html');
  let pathToFiles;
  const localFiles = path.resolve(pathDir, dirName);
  return html.then(() => fs.writeFile(path.resolve(pathDir, pageName), page))
    .then(() => fs.mkdir(localFiles))
    .then(() => {
      pathToFiles = localFiles;
      return pathToFiles;
    })
    .then(() => loadAndSaveData(promises, paths, pathToFiles));
};

export default pageLoader;
