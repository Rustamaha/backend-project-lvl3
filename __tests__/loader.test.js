import nock from 'nock';
import os from 'os';

import path from 'path';
import { promises as fs } from 'fs';

import pageLoader from '../src';
import { makeName } from '../src/loader';

const debug = require('debug');

const log = debug('page-loader:test');

const url = 'https://ru.hexlet.io/courses';

const removeDirWithFiles = async (pathDir) => {
  const data = await fs.readdir(pathDir);
  if (data.length === 0) {
    await fs.rmdir(pathDir);
  } else {
    await Promise.all(data.map(async (file) => {
      const pathSubData = path.join(pathDir, file);
      const stats = await fs.stat(pathSubData);
      if (stats.isDirectory()) {
        await removeDirWithFiles(pathSubData);
      } else {
        await fs.unlink(pathSubData);
      }
    }));
    await fs.rmdir(pathDir);
  }
};

const buildPathForFixture = (format, fileName) => path.join('__tests__', '__fixtures__', `${fileName}${format}`);
const buildPathForHexletFiles = (format, fileName) => path.join('__tests__', '__fixtures__', 'ru-hexlet-io-courses_files', `${fileName}${format}`);
const originalHexletPage = buildPathForFixture('.html', 'origin-hexlet-page');
const convertedHexletPage = buildPathForFixture('.html', 'ru-hexlet-io-courses');
const cssFile1 = buildPathForHexletFiles('.css', 'cdn2-hexlet-io-packs-css-36-e1004b13-chunk-css');
const cssFile2 = buildPathForHexletFiles('.css', 'cdn2-hexlet-io-assets-application-58b8be69d43878d8ffa548a26a341422323098508999ea2cd5f001896ad189dc-css');
const rssFile = buildPathForHexletFiles('.rss', 'ru-hexlet-io-lessons-rss');

let tempDir;
let hexletPagePath;
let cssFilePath1;
let cssFilePath2;
let rssFilePath;
let expectedHexletPage;
let expectedCssFile1;
let expectedCssFile2;
let expectedRssFile;
const hexletFile = makeName(url, '.html');
const localFiles = makeName(url, '_files');
const cssFileName1 = 'cdn2-hexlet-io-packs-css-36-e1004b13-chunk-css.css';
const cssFileName2 = 'cdn2-hexlet-io-assets-application-58b8be69d43878d8ffa548a26a341422323098508999ea2cd5f001896ad189dc-css.css';
const rssFileName = 'ru-hexlet-io-lessons-rss.rss';

beforeEach(async () => {
  const temp = `${os.tmpdir()}${path.sep}`;
  log(temp);
  const tmpDir = await fs.mkdtemp(temp);
  tempDir = tmpDir;
  log(`tempDir ${tmpDir}`);
  hexletPagePath = path.resolve(tmpDir, hexletFile);
  cssFilePath1 = path.resolve(tmpDir, localFiles, cssFileName1);
  cssFilePath2 = path.resolve(tmpDir, localFiles, cssFileName2);
  rssFilePath = path.resolve(tmpDir, localFiles, rssFileName);
  expectedHexletPage = await fs.readFile(convertedHexletPage, 'utf8');
  expectedCssFile1 = await fs.readFile(cssFile1, 'utf8');
  expectedCssFile2 = await fs.readFile(cssFile2, 'utf8');
  expectedRssFile = await fs.readFile(rssFile, 'utf8');
});

afterEach(async () => {
  await removeDirWithFiles(tempDir);
  try {
    await fs.readdir(tempDir);
  } catch (err) {
    log(`rmdir tempDir path: ${err.path}`);
  }
});

test(`a page from ${url} is successfully converted and saved local files`, async () => {
  await nock('https://ru.hexlet.io')
    .get('/courses')
    .reply(200, () => fs.readFile(originalHexletPage, 'utf8'));
  nock('https://cdn2.hexlet.io')
    .get('/assets/application-6c3811f32b2b06662856f28be1aa8852645cc103fce8b59a6a05e08ae031ee55.js')
    .reply(200, {
      data: 'hello world',
    });
  nock('https://cdn2.hexlet.io')
    .get('/packs/css/36-e1004b13.chunk.css')
    .reply(200, () => fs.readFile(cssFile1, 'utf8'));
  nock('https://ru.hexlet.io')
    .get('/cdn-cgi/scripts/5c5dd728/cloudflare-static/email-decode.min.js')
    .reply(200, {
      data: 'hello world',
    });
  nock('https://cdn2.hexlet.io')
    .get('/assets/application-58b8be69d43878d8ffa548a26a341422323098508999ea2cd5f001896ad189dc.css')
    .reply(200, () => fs.readFile(cssFile2, 'utf8'));
  nock('https://ru.hexlet.io')
    .get('/lessons.rss')
    .reply(200, () => fs.readFile(rssFile, 'utf8'));
  nock('https://ru.hexlet.io')
    .get('/courses')
    .reply(200, {
      data: 'hello world',
    });

  await pageLoader(url, tempDir);

  const data1 = await fs.readFile(hexletPagePath, 'utf8');
  expect(data1).toBe(expectedHexletPage);
  const data2 = await fs.readFile(cssFilePath1, 'utf8');
  expect(data2).toBe(expectedCssFile1);
  const data3 = await fs.readFile(cssFilePath2, 'utf8');
  expect(data3).toBe(expectedCssFile2);
  const data4 = await fs.readFile(rssFilePath, 'utf8');
  expect(data4).toBe(expectedRssFile);
});

describe('some network problems', () => {
  test(`a page from ${url} reply with status code: 404`, async () => {
    nock('https://ru.hexlet.io')
      .get('/courses')
      .reply(404);

    await expect(pageLoader(url, tempDir))
      .rejects
      .toThrowErrorMatchingSnapshot();
  });

  test(`a page from ${url} reply with status code 403`, async () => {
    await nock('https://ru.hexlet.io')
      .get('/courses')
      .reply(403);

    await expect(pageLoader(url, tempDir))
      .rejects
      .toThrowErrorMatchingSnapshot();
  });

  test(`a page from ${url} reply with status 200, but some files from link reply with status 404`, async () => {
    await nock('https://ru.hexlet.io')
      .get('/courses')
      .reply(200, () => fs.readFile(originalHexletPage, 'utf8'));
    nock('https://cdn2.hexlet.io')
      .get('/assets/application-6c3811f32b2b06662856f28be1aa8852645cc103fce8b59a6a05e08ae031ee55.js')
      .reply(200, {
        data: 'hello world',
      });
    nock('https://cdn2.hexlet.io')
      .get('/packs/css/36-e1004b13.chunk.css')
      .reply(200, () => fs.readFile(cssFile1, 'utf8'));
    nock('https://ru.hexlet.io')
      .get('/cdn-cgi/scripts/5c5dd728/cloudflare-static/email-decode.min.js')
      .reply(200, {
        data: 'hello world',
      });
    nock('https://cdn2.hexlet.io')
      .get('/assets/application-58b8be69d43878d8ffa548a26a341422323098508999ea2cd5f001896ad189dc.css')
      .delay(5000)
      .reply(404);
    nock('https://ru.hexlet.io')
      .get('/lessons.rss')
      .reply(200, () => fs.readFile(rssFile, 'utf8'));
    nock('https://ru.hexlet.io')
      .get('/courses')
      .reply(200, {
        data: 'hello world',
      });

    await pageLoader(url, tempDir);

    const data1 = await fs.readFile(hexletPagePath, 'utf8');
    expect(data1).toBe(expectedHexletPage);
    const data2 = await fs.readFile(cssFilePath1, 'utf8');
    expect(data2).toBe(expectedCssFile1);
    const data3 = await fs.readFile(rssFilePath, 'utf8');
    expect(data3).toBe(expectedRssFile);
  }, 10000);
});

describe('some file system problems', () => {
  test('called without any directory', async () => {
    await expect(pageLoader(url, ''))
      .rejects
      .toThrowErrorMatchingSnapshot();
  });

  test('called with non-existent directory', async () => {
    await expect(pageLoader(url, '/qwerty'))
      .rejects
      .toThrowErrorMatchingSnapshot();
  });
});
