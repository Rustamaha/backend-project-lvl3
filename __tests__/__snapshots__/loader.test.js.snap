// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`some file system problems called with non-existent directory 1`] = `"ENOENT: no such file or directory, stat '/qwerty'"`;

exports[`some file system problems called without any directory 1`] = `"ENOENT: no such file or directory, stat ''"`;

exports[`some network problems a page from https://ru.hexlet.io/courses reply with status code 403 1`] = `"Request failed with status code 403"`;

exports[`some network problems a page from https://ru.hexlet.io/courses reply with status code: 404 1`] = `"Request failed with status code 404"`;
